crates/factor-outbound-networking/src/runtime_config.rs

```diff
@@ -0,0 +1,250 @@
+}
+
+impl ComponentTlsConfigs {
+    /// Returns a [`ClientConfig`] for the given host authority.
+    pub fn get_client_config(&self, host: &str) -> &Arc<ClientConfig> {
+        self.host_client_configs
+            .as_ref()
+            .and_then(|configs| configs.get(host))
+            .unwrap_or(&self.default_client_config)
+    }
+}
+
+#[derive(Debug)]
+pub struct ClientCertConfig {
+    cert_chain: Vec<CertificateDer<'static>>,
+    key_der: PrivateKeyDer<'static>,
+}
+
+#[derive(Debug)]
+pub struct TlsConfig {
+    pub components: Vec<String>,
+    pub hosts: Vec<String>,
+    pub root_certificates: Vec<rustls_pki_types::CertificateDer<'static>>,
+    pub use_webpki_roots: bool,
+    pub client_cert: Option<ClientCertConfig>,
+}
+
+impl Default for TlsConfig {
+    fn default() -> Self {
+        Self {
+            components: vec![],
+            hosts: vec![],
+            root_certificates: vec![],
+            // Use webpki roots by default
+            use_webpki_roots: true,
+            client_cert: None,
+        }
+    }
+}
+
+impl TlsConfig {
+    fn to_client_config(&self) -> anyhow::Result<ClientConfig> {
+        let mut root_store = RootCertStore::empty();
+        if self.use_webpki_roots {
+            root_store.extend(webpki_roots::TLS_SERVER_ROOTS.iter().cloned());
+        }
+        for ca in &self.root_certificates {
+            root_store.add(ca.clone())?;
+        }
+
+        let builder = ClientConfig::builder().with_root_certificates(root_store);
+
+        if let Some(ClientCertConfig {
+            cert_chain,
+            key_der,
+        }) = &self.client_cert
+        {
+            Ok(builder.with_client_auth_cert(cert_chain.clone(), key_der.clone_key())?)
+        } else {
+            Ok(builder.with_no_client_auth())
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use std::{io::BufReader, path::Path};
+
+    use anyhow::Context;
+
+    use super::*;
+
+    #[test]
+    fn test_empty_config() -> anyhow::Result<()> {
+        let runtime_config = RuntimeConfig::new([])?;
+        // Just make sure the default path doesn't panic
+        runtime_config.get_client_config("foo", "bar");
+        Ok(())
+    }
+
+    #[test]
+    fn test_minimal_config() -> anyhow::Result<()> {
+        let runtime_config = RuntimeConfig::new([TlsConfig {
+            components: vec!["test-component".into()],
+            hosts: vec!["test-host".into()],
+            root_certificates: vec![],
+            use_webpki_roots: false,
+            client_cert: None,
+        }])?;
+        let client_config = runtime_config.get_client_config("test-component", "test-host");
+        // Check that we didn't just get the default
+        let default_config = runtime_config.get_client_config("other_component", "test-host");
+        assert!(!Arc::ptr_eq(&client_config, &default_config));
+        Ok(())
+    }
+
+    #[test]
+    fn test_maximal_config() -> anyhow::Result<()> {
```

Comment:
Nit: I'm not sure this a maximal config considering `use_webpki_roots` is set to false?

```suggestion
    fn test_custom_client_cert() -> anyhow::Result<()> {
```

Reply:
It is "maximally different from the defaults" :slightly_smiling_face: